File: 02-encapsulamiento/python/cuenta_bancaria.py
"""
cuenta_bancaria.py - LA CLASE

Nombre del archivo en minusculas con guion bajo, clase en CamelCase.
Al importarla se vera asi:
    from cuenta_bancaria import CuentaBancaria

LA GRAN DIFERENCIA CON JAVA
  Java tiene privacidad REAL: el compilador te impide tocar un private.
  Python tiene privacidad POR ACUERDO: se avisa, no se prohibe.

  Convenciones de Python:
    saldo      -> publico
    _saldo     -> "no lo toques, es interno" (solo una senal, nada lo impide)
    __saldo    -> name mangling: se renombra a _Clase__saldo, molesta de verdad
                  pero sigue siendo accesible si te empenas

  Y en vez de getSaldo()/setSaldo(), Python usa @property: se escribe como
  un atributo pero por dentro ejecuta un metodo con validacion.
"""


class CuentaBancaria:

    def __init__(self, titular, saldo_inicial):
        self._titular = titular
        self.__saldo = max(saldo_inicial, 0)   # __ = lo mas parecido a private

    # ---------- GETTER estilo Python ----------
    # Se usa como cuenta.saldo (SIN parentesis), pero ejecuta este metodo.
    @property
    def saldo(self):
        return self.__saldo

    # No definimos un setter para saldo -> queda de SOLO LECTURA.

    @property
    def titular(self):
        return self._titular

    # ---------- SETTER estilo Python, con validacion ----------
    @titular.setter
    def titular(self, nuevo_titular):
        if not nuevo_titular:
            print("ERROR: el titular no puede estar vacio.")
            return
        self._titular = nuevo_titular

    # ---------- Reglas de negocio ----------
    def depositar(self, monto):
        if monto <= 0:
            print("ERROR: el deposito debe ser positivo.")
            return
        self.__saldo += monto
        print(f"Deposito de {monto} OK.")

    def girar(self, monto):
        if monto > self.__saldo:
            print(f"ERROR: saldo insuficiente para girar {monto}")
            return
        self.__saldo -= monto
        print(f"Giro de {monto} OK.")

    def mostrar(self):
        print(f"[Cuenta] {self._titular} | saldo: ${self.__saldo}")

File: 02-encapsulamiento/java/CuentaBancaria.java
/*
 * CuentaBancaria.java - LA CLASE
 *
 * Ahora que la clase vive en su PROPIO archivo se ve mejor de que sirve
 * "private": es la frontera entre lo que este archivo controla y lo que
 * el resto del programa puede tocar.
 *
 * Un atributo public es como dejar la caja registradora abierta.
 * Un atributo private obliga a pasar por el cajero (los metodos).
 */
public class CuentaBancaria {

    // PRIVATE = solo el codigo de ESTE archivo puede tocar estos atributos
    private String titular;
    private int saldo;

    public CuentaBancaria(String titular, int saldoInicial) {
        this.titular = titular;
        this.saldo = Math.max(saldoInicial, 0);   // nunca nace negativa
    }

    // ---------- GETTERS: dejan LEER ----------
    public String getTitular() {
        return this.titular;
    }

    public int getSaldo() {
        return this.saldo;
    }

    // ---------- SETTER: deja ESCRIBIR, pero con reglas ----------
    public void setTitular(String nuevoTitular) {
        if (nuevoTitular == null || nuevoTitular.isEmpty()) {
            System.out.println("ERROR: el titular no puede estar vacio.");
            return;
        }
        this.titular = nuevoTitular;
    }

    // Fijate que NO existe setSaldo(). El saldo solo cambia con reglas
    // de negocio: depositar y girar. Eso es encapsular de verdad.

    public void depositar(int monto) {
        if (monto <= 0) {
            System.out.println("ERROR: el deposito debe ser positivo.");
            return;
        }
        this.saldo += monto;
        System.out.println("Deposito de " + monto + " OK.");
    }

    public void girar(int monto) {
        if (monto > this.saldo) {
            System.out.println("ERROR: saldo insuficiente para girar " + monto);
            return;
        }
        this.saldo -= monto;
        System.out.println("Giro de " + monto + " OK.");
    }

    public void mostrar() {
        System.out.println("[Cuenta] " + this.titular + " | saldo: $" + this.saldo);
    }
}
